### PR TITLE
File Settings Service

### PR DESCRIPTION
**Pull request by grcevski**
Watches a directory under Elasticsearch config for changes to a file called settings.json. This is a JSON file describing the settings that the operator wants to set that will be immutable through the REST API. Any changes to these settings will only be allowed through modifying the file. As configured by default we watch for a path operator/settings.json under the config directory. The name of the directory operator can be changed via settings.

The File Settings Service will notice updates to the file and perform an update. The operator directory doesn't have to exist, the file setting service will notice when it's created. Recommended way to change operator file based settings is through modifying the settings.json file in a brand new location and then performing a symlink under the Elasticsearch config directory.

Relates to https://github.com/elastic/elasticsearch/pull/86224

### CODE REVIEW

**Comment by grcevski** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:159]: trivial: not a javadoc, so no need for the extra `*`

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:176]: nit: This seems more like a debug level message?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:169]: nit: This seems more like a debug level message?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:177]: If stashing the expected operator path as a member instead of environment, then this can be `operatorSettingsDir.getParent()`

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:221]: Is there any cost or potential negative behavior from calling register after every take()? ie, why not only do this when the key indicates the parent dir was created?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:276]: nit: maybe do some static imports and/or set the json xcontent so that the try can be on one line (for readability)

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:283]: It looks like this gets propagated because the exception consumer is not guarded. That's fine, but then we catch the exception down below, so it seems like we would not fail startup as intended? Am I missing something?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:227]: Why the "watch" part of this message? There shouldn't even be an exception out of asking to watch the settings dir right?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:290]: process doesn't throw anything checked. So why is this guarded? There is already a catch outside of this method that should handled logging the exception?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:238]: This would only happen when an exception occurs, so is this useful as an info message? Why not log the exception at error level, stating the watcher thread is shutting down? Isn't this always an exception case (not how the thread would be shutdown)?

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:254]: why is this info instead of error or warn?

File: server/src/test/java/org/elasticsearch/action/ActionModuleTests.java
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/action/ActionModuleTests.java:124]: trivial: all these `Collections.emptyList()` can be `List.of()`

File: test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java
Comment (rjernst) [inline test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java:85]: Can this be moved out to a separate PR, with a unit test?

**Review by rjernst — COMMENTED** [review]
Looks pretty good! I have several comments, mostly nits

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (ChrisHegarty) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:276]: Unless I'm mistaken we do not set `StreamReadFeature.AUTO_CLOSE_SOURCE`, so we need to explicitly close the input stream returned from Files::newInputStream, right? Maybe just refactor into the first declaration in the try-with-resources, then pass to createParser. And maybe wrap in a BufferedInputStream ?

File: test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java
Comment (grcevski) [reply test/framework/src/main/java/org/elasticsearch/plugins/MockPluginsService.java:85]: Sounds good, I made a new PR here https://github.com/elastic/elasticsearch/pull/88690. Once that's merged I'll rebase and remove this change.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:254]: I'll switch to warn.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:238]: I didn't want to pollute the logs with an exception trace when the watching is interrupted on the take() method call. It's how we interrupt the watcher when we stop watching, e.g. we are no longer the master node. So what I'll do here is log info when ClosedWatchServiceException is thrown and log error on any other exception.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:227]: Good point, I'll make it generic, it's a remnant of a refactor I had done.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:283]: OperatorConfigurationError is of type Error, so it's never caught. I wasn't sure if there's a nicer way to cause the exception to skip all the exception handling we do and terminate the node.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:221]: I'll double check to see what happens when the watcher is over and over registered on the same dir, I didn't see an issue when testing, but it doesn't mean it doesn't leak somehow. 

The reason I enable the settings watcher every time is because of symlinks. It's platform specific whether the watcher continues to watch the original symlinked directory or the new one after a change to a link. I couldn't think of another way to make this more reliable, other than repeat that I want to watch the same dir.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:238]: We could have a separate catch for InterruptedException so that any unknown errors are logged at the appropriate level?

File: server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java
Comment (grcevski) [reply server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java:87]: I agree that's this is tedious to modify each time. We need to explicitly add each thread that's not in any threadpool. When I worked on the readiness-service, I was advised not to use the thread pools for threads that will constantly be running, so I followed the same approach here. `FileSystemWatchService` comes from the JDK, so there's no avoiding adding that one.

Is it OK if I followed up with a new PR to tackle this problem, rather than extend this PR further? I think I should be able to ensure all standalone Elasticsearch threads are registered in a way that this list doesn't have to be manually modified.

**Comment by grcevski** [comment]
I think I addressed all of the feedback and I added an integration test to ensure it all is wired up together and it's working.

File: server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java
Comment (grcevski) [inline server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java:86]: I will tackle cleaning up this manual registration in a separate PR.

File: server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java
Comment (rjernst) [reply server/src/internalClusterTest/java/org/elasticsearch/threadpool/SimpleThreadPoolIT.java:87]: A followup is great

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:273]: Stop happens on another ES thread, not one controlled here. So the interrupt should be restored for InterruptedException. (perhaps capture InterruptedException in another catch block). Note I think it is ok to not restore above since the thread is owned by this service and the thread will end after the interrupted exception is logged there.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:243]: How is this possible given processFileSettings logs any errors? It seems like the only possible error here is IOException? Could this be narrowed to that exception, and the warning message be distinguished from that on line 307? They are very similarly worded, yet this one, I don't think is specific to processing the file, it's just some IO problem happened while trying read it.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:325]: We shouldn't be extending Error (and should be renamed to avoid configuration, perhaps mentioning this is a startup problem in the name). This can extend RuntimeException so it can be propagated easily.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:304]: The logic here is a little difficult to reason about since it depends on this boolean (but the error here is then not possible outside of startup). Perhaps a more direct approach would be to pass a Consumer<Exception> so the two callers can decide how to handle it.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:66]: Why is this configurable? I think we could start with it hardcoded? I much prefer it is always the same, so that we know it will always be directly under the config dir.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:171]: Rather than grab the default filesystem, it is probably better to grab it from the config dir?

**Review by rjernst — APPROVED** [review]
A few more comments. LGTM, no need for anymore review if you agree with them.

File: server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java
Comment (grcevski) [reply server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java:66]: Sure thing, I'll fix it.

**Pull request merged**
